src/bson.rs

```diff
@@ -973,76 +970,78 @@ impl Timestamp {
+///     chrono_datetime: chrono::DateTime<chrono::Utc>,
+///
+///     // serializes as a BSON datetime.
+///     #[serde(with = "bson::serde_helpers::chrono_0_4_datetime_as_bson_datetime")]
+///     chrono_as_bson: chrono::DateTime<chrono::Utc>,
 /// }
 /// ```
 #[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Copy, Clone)]
-pub struct DateTime(pub chrono::DateTime<Utc>);
-
-impl DateTime {
-    pub(crate) fn from_i64(date: i64) -> Self {
-        let mut num_secs = date / 1000;
-        let mut num_millis = date % 1000;
-
-        // The chrono API only lets us create a DateTime with an i64 number of seconds
-        // and a u32 number of nanoseconds. In the case of a negative timestamp, this
-        // means that we need to turn the negative fractional part into a positive and
-        // shift the number of seconds down. For example:
-        //
-        //     date       = -4300 ms
-        //     num_secs   = date / 1000 = -4300 / 1000 = -4
-        //     num_millis = date % 1000 = -4300 % 1000 = -300
-        //
-        // Since num_millis is less than 0:
-        //     num_secs   = num_secs -1 = -4 - 1 = -5
-        //     num_millis = num_nanos + 1000 = -300 + 1000 = 700
-        //
-        // Instead of -4 seconds and -300 milliseconds, we now have -5 seconds and +700
-        // milliseconds, which expresses the same timestamp, but in a way we can create
-        // a DateTime with.
-        if num_millis < 0 {
-            num_secs -= 1;
-            num_millis += 1000;
-        };
+pub struct DateTime(chrono::DateTime<Utc>);
 
-        Utc.timestamp(num_secs, num_millis as u32 * 1_000_000)
-            .into()
+impl crate::DateTime {
+    pub(crate) fn from_millis(date: i64) -> Self {
+        Utc.timestamp_millis(date).into()
     }
-}
 
-impl Deref for DateTime {
-    type Target = chrono::DateTime<Utc>;
+    /// Returns the number of non-leap-milliseconds since January 1, 1970 UTC.
+    pub fn timestamp_millis(&self) -> i64 {
+        self.0.timestamp_millis()
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+    /// If this DateTime is sub-millisecond precision. BSON only supports millisecond precision, so
+    /// this should be checked before serializing to BSON.
+    pub(crate) fn is_sub_millis_precision(&self) -> bool {
+        self.0.timestamp_subsec_micros() % 1000 != 0
     }
 }
 
-impl DerefMut for DateTime {
-    fn deref_mut(&mut self) -> &mut chrono::DateTime<Utc> {
-        &mut self.0
+impl Display for crate::DateTime {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.0, f)
     }
 }
 
-impl From<DateTime> for chrono::DateTime<Utc> {
+impl From<crate::DateTime> for chrono::DateTime<Utc> {
     fn from(utc: DateTime) -> Self {
         utc.0
     }
 }
 
-impl From<chrono::DateTime<Utc>> for DateTime {
-    fn from(x: chrono::DateTime<Utc>) -> Self {
-        DateTime(x)
+impl<T: chrono::TimeZone> From<chrono::DateTime<T>> for crate::DateTime {
```

Comment:
I updated this to be more general since it could be a little annoying to convert from `DateTime`s with other timezones otherwise.